scripts/irods/test/test_prep_genquery_iterator.py

```diff
@@ -832,3 +832,117 @@ def tearDown(self):
 
         super(Test_Genquery_Iterator, self).tearDown()
 
+    @unittest.skipUnless(plugin_name == 'irods_rule_engine_plugin-python', 'Requires PREP.')
+    def test_genquery_constructor_raises_exception_on_invalid_parser_argument__issue_7909(self):
+        with temporary_core_file() as core:
+            attr_name = 'test_genquery_constructor_raises_exception_on_invalid_parser_argument__issue_7909_error'
+
+            # The following rule triggers and attaches the exception message generated
+            # by the invalid parser argument to the Query constructor.
+            core.add_rule(dedent(f'''
+            def pep_api_touch_pre(rule_args, callback, rei):
+                try:
+                    import genquery
+                    _ = genquery.Query(callback, 'COLL_NAME', parser=None)
+                except ValueError as e:
+                    callback.msiModAVUMetadata('-C', '{self.admin.session_collection}', 'add', '{attr_name}', str(e), '')
+            '''))
+
+            # Trigger the PEP.
+            # This will cause the rule to add an AVU to the session collection.
+            self.admin.assert_icommand(['itouch', 'issue_7909'], 'STDERR')
+
+            # Show the AVU containing the error exists on the session collection.
+            expected_output = [
+                f'attribute: {attr_name}\n',
+                'value: Invalid value for [parser]. Expected Parser.GENQUERY1 or Parser.GENQUERY2.\n'
+            ]
+            self.admin.assert_icommand(['imeta', 'ls', '-C', self.admin.session_collection], 'STDOUT', expected_output)
+
+    @unittest.skipUnless(plugin_name == 'irods_rule_engine_plugin-python', 'Requires PREP.')
+    def test_genquery_iterator_supports_genquery2__issue_7909(self):
+        rule_file = f'{self.admin.local_session_dir}/test_genquery_iterator_supports_genquery2__issue_7909.r'
+
+        with open(rule_file, 'w') as rf:
+            rf.write(dedent(f'''
+            def main(rule_args, callback, rei):
+                from genquery import Query, Parser
+                for r in Query(callback, 'COLL_NAME', "COLL_NAME = '{self.admin.session_collection}'", parser=Parser.GENQUERY2):
```

Comment:
Any value in trying to use some GenQuery2-only features in these tests?

Should we add a test which attempts to use GenQuery2-only features with the `Parser.GENQUERY1` parser? Not for the sake of testing the parser, but more to make sure the right parser is being used (in this case, by watching it break when the syntax for the other parser is in use).

Maybe there isn't any value to that.

Reply:
If the GenQuery2 interface was different, I think it would be good to test that. But since GenQuery2 just takes a single string, I don't think there's much value in testing that.

Most of the GenQuery2 logic in the module just ignores GenQuery1 stuff.

With that said, do you see value in adding a test for that? What would the test do?

Reply:
After some discussion, we've agreed that adding a test to verify the correct parser is being used has value.

Reply:
New test looks good